### PR TITLE
How borrowedBooks.Add((BorrowedBook)book);

### PR DESCRIPTION
**Pull request by prajdziswet**

### CODE REVIEW

**Comment by prajdziswet** [comment]
что то я плаваю с наследованием, хоть это базовый принцип...наверное не так часто наследовал...

File: ClassLibrary/ClassLibrary/BorrowedBook.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        public Book book
+        {
+            get;
+            private set;
+        }
+
+        private DateTime _dateTime;
+        public DateTime dateTime
+        {
+            get => _dateTime;
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/BorrowedBook.cs]: why not just a private setter?

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs [hunk 2]
```diff
@@ -82,5 +82,48 @@ public void BorrowedBookNullArgument()
             Should.Throw<ArgumentNullException>(() => DR.BorrowBook(reader, null));
         }
 
+        [Test]
+        public void NotList_GetTimeWhenFreeBook()
+        {
+            DepartmentReaders DR = new DepartmentReaders();
+            //I think it doesn't matter which parameter returns. What do you think?
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs]: you're right, there's no "right" answer. but you should be able to make a grounded choice, so think what are pros and cons of each option.

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs [hunk 3]
```diff
@@ -82,5 +82,48 @@ public void BorrowedBookNullArgument()
...
+            Should.Throw<ArgumentNullException>(() => DR.GetTimeWhenFreeBook(null));
+            Should.Throw<ArgumentNullException>(() => DR.GetTimeWhenFreeBook(new System.Collections.Generic.List<BorrowedBook>()));
+        }
+
+        [Test]
+        public void ReturnNotExistBook()
+        {
+            DepartmentReaders DR = new DepartmentReaders();
+            Author author = new Author("Lev", "Tolstoj");
+            Book book = new Book("226611156", "War and Peace", author);
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs:100]: you don't need to create `author` and `book` here

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs [hunk 4]
```diff
@@ -82,5 +82,48 @@ public void BorrowedBookNullArgument()
...
+            DR.AddReader(reader);
+
+            Should.Throw<ArgumentException>(() => DR.ReturnBook(reader.ID, "Mu-Mu")).Message.ShouldBe("you didn't take this book");
+        }
+
+        [Test]
+        public void ReturnNotSetBook()
+        {
+            DepartmentReaders DR = new DepartmentReaders();
+            Author author = new Author("Lev", "Tolstoj");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs]: you don't need to create `author` and `book` here

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs [hunk 5]
```diff
@@ -82,5 +82,48 @@ public void BorrowedBookNullArgument()
...
+            DR.AddReader(reader);
+
+            Should.Throw<ArgumentNullException>(() => DR.ReturnBook(reader.ID, "")).ParamName.ShouldBe("NameBook");
+        }
+
+        [Test]
+        public void ReturnBookNotExistReader()
+        {
+            DepartmentReaders DR = new DepartmentReaders();
+            Author author = new Author("Lev", "Tolstoj");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs:136]: you don't need to create `author`, `book` and `reader` here

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs [hunk 6]
```diff
@@ -82,5 +82,48 @@ public void BorrowedBookNullArgument()
             Should.Throw<ArgumentNullException>(() => DR.BorrowBook(reader, null));
         }
 
+        [Test]
+        public void NotList_GetTimeWhenFreeBook()
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs:87]: I expected a positive case (actually, a lot of positive cases) here. why there are none? it's a complex error-prone logic.

File: ClassLibrary/ClassLibrary.Test/TestEntities.cs [hunk 7]
```diff
@@ -141,5 +140,33 @@ public void AddEqualBooksInCard()
 
             Should.Throw<ArgumentException>(() => DR.BorrowBook(reader, book));
         }
+
+        [Test]
+        public void DeleteBookInCard()
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestEntities.cs:145]: why this test is not in DepartmentReaders test class?

File: ClassLibrary/ClassLibrary.Test/TestEntities.cs [hunk 8]
```diff
@@ -141,5 +140,33 @@ public void AddEqualBooksInCard()
...
+        {
+            Reader reader = new Reader("Lev", "Tolstoj");
+            Author author = new Author("Lev", "Tolstoj");
+            Book book = new Book("226611156", "War and Peace", author);
+            Book book1 = new Book("226611150", "Mu-Mu", author);
+            DepartmentReaders DR = new DepartmentReaders();
+            DR.AddReader(reader);
+            DR.BorrowBook(reader, book);
+
+            Should.Throw<ArgumentException>(() => DR.ReturnBook(reader.ID, book1.NameBook)).Message.ShouldBe("you didn't take this book");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestEntities.cs]: You don't need to create `book1` when all you need is a name.

File: ClassLibrary/ClassLibrary.Test/TestLibrary.cs [hunk 9]
```diff
@@ -40,6 +39,8 @@ public void NotReader()
 
             Reader reader = new Reader("Ivan", "Ivanov");
             Should.Throw<ArgumentException>(() => lib.ReaderBoroweBook(reader.ID, book.NameBook)).Message.ShouldBe($"Not Exist Reader with (ID={reader.ID}) in DepartmentReaders");
+            //for return book
+            Should.Throw<ArgumentException>(() => lib.ReaderReturnBook(reader.ID, book.NameBook)).Message.ShouldBe($"Not Exist Reader with (ID={reader.ID}) in DepartmentReaders");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestLibrary.cs]: I don't think it's the best practice to test different methods in the same test, but it's ok as long as you can tell what are pros and cons of both approaches.

File: ClassLibrary/ClassLibrary.Test/TestLibrary.cs [hunk 10]
```diff
@@ -92,7 +95,7 @@ public void ALLBooKWithISBNBorrowed()
             lib.ReaderBoroweBook(reader1.ID, book1.NameBook);
             lib.ReaderBoroweBook(reader2.ID, book2.NameBook);
 
-            Should.Throw<ArgumentException>(() => lib.ReaderBoroweBook(reader3.ID, book1.NameBook)).Message.ShouldBe("ALL books borrowed");
+            Should.Throw<ArgumentException>(() => lib.ReaderBoroweBook(reader3.ID, book1.NameBook));
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary.Test/TestLibrary.cs]: Why did you remove the message?

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs [hunk 11]
```diff
@@ -48,9 +48,20 @@ public int CountBorrowedBooksWithISBN(String ISBN)
...
         }
 
-        public HashSet<Book> BorrowedBooksWithISBN(String ISBN)
+        public List<BorrowedBook> BorrowedBooksWithISBN(String ISBN)
         {
-            return new HashSet<Book>(Readers.SelectMany(x => x.BorrowedBooks.Where(y => y.ISBN == ISBN)));
+            return Readers.SelectMany(x => x.BorrowedBooks.Where(y => y.book.ISBN == ISBN)).ToList();
+        }
+
+        public TimeSpan GetTimeWhenFreeBook(List<BorrowedBook> borrowedBooksISBN)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/DepartmentReaders.cs]: The return type should be `DateTime`, not `TimeSpan`. it's much easier to understand that "the book is available on June 23rd" than "the book is available in 16 days, 18 hours and 44 minutes".

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs [hunk 12]
```diff
@@ -48,9 +48,20 @@ public int CountBorrowedBooksWithISBN(String ISBN)
...
+        {
+            if (borrowedBooksISBN==null||borrowedBooksISBN.Count==0)
+            {
+                throw new ArgumentNullException(nameof(borrowedBooksISBN), "In DepartmentReaders.GetTimeWhenFreeBook");
+            }
+            DateTime timeNow = DateTime.Now;
+            BorrowedBook moreBestFreeBook = borrowedBooksISBN.OrderBy(book => book.dateTime).FirstOrDefault(book => (timeNow - book.dateTime).Days < 30);
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/DepartmentReaders.cs]: when I read this code I realize that if someone didn't return the book in 30 days it just "disappears". if that's intentional, to not confuse users with the response "will be available yesterday" then there must be a test for that specific case.
without a test case, it just looks like a coding mistake.

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs [hunk 13]
```diff
@@ -81,5 +92,36 @@ public void BorrowBook(Reader reader,Book book)
 
             reader.AddBookInCard(book);
         }
+
+        public void ReturnBook(int IDReader, String NameBook)
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/DepartmentReaders.cs]: I understand why this method has the same parameters as `BorrowBook`, but it can't be the same.
imagine real life situations: 
1. reader comes to the librarian and says "I need War and Peace" - all librarian has to fulfill that request is reader's card with ID and a book name.
2. reader comes to the librarian and says "I want to return this War and Peace book" - librarian doesn't need to check look for a book by its name - the book is already there, it has a tag on it.

so there should be just book id instead of book name.

File: ClassLibrary/ClassLibrary/Library.cs [hunk 14]
```diff
@@ -43,24 +43,24 @@ public void ReaderBoroweBook(int IDReader, String NameBook)
...
             //Free Book
-            allBookWithISBN.ExceptWith(borrowedBookWithISBN);
+            List<Book> freebooks = listAllBookWithName.FreeBook(borrowedBookWithISBN);
 
-            if (allBookWithISBN.Count==0)
+            if (freebooks.Count==0)
             {
-                throw new ArgumentException("ALL books borrowed");
+                TimeSpan days=departmentReaders.GetTimeWhenFreeBook(borrowedBookWithISBN);
+                throw new ArgumentException($"ALL books borrowed. {days.Days} days when the nearest book is free ");
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/Library.cs]: "all books borrowed" is not an exceptional situation according to the requirements, you shouldn't throw an exception here.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary/Library.cs (hunk 14 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: corrected in new version

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 13 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: not done yet

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 12 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: пока пропустил, это сделано намеренно 
=> в библиотеке должны решать, что делать с должником //по хорошему да надо добавить теэст

File: ClassLibrary/ClassLibrary/BorrowedBook.cs (hunk 1 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/BorrowedBook.cs]: correct

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs (hunk 2 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs]: +

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs (hunk 3 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs:100]: +

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs (hunk 4 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs]: +

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs (hunk 5 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs:136]: +

File: ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs (hunk 6 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestDepartmentReaders.cs:87]: добавил положительный случай

File: ClassLibrary/ClassLibrary.Test/TestEntities.cs (hunk 7 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestEntities.cs:145]: I prefer being closer to the essence(entity) and not duplicating code. Although sometimes I can contact through DepartmentReaders, which is not very pleasant =(

File: ClassLibrary/ClassLibrary.Test/TestEntities.cs (hunk 8 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestEntities.cs]: +

File: ClassLibrary/ClassLibrary.Test/TestLibrary.cs (hunk 9 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestLibrary.cs]: - проверка, что оба метода возвращают одно и то же сообщения = одинаковые ошибки в двух методах
- название метода-теста более понятно
- не дублируеться одинаковые ошибки
"-" сложнее понять где произощло именно падение теста

File: ClassLibrary/ClassLibrary.Test/TestLibrary.cs (hunk 10 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestLibrary.cs]: 1) Я думаю, что сообщение лучше в UI а не тесте
2) возвращает сообщение с датой

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 11 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: ok

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 13 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: I have corrected, but it may be like this. The librarian will not check by code (if there is no scanner), if it contains 10 digits, it is easier for him to check by name. And I doubt that you will take 50 books

File: ClassLibrary/ClassLibrary/Library.cs (hunk 14 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: there is a check in the code, but I cannot cover it with a test

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 12 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: добавил тест

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs [hunk 15]
```diff
@@ -53,15 +53,16 @@ public List<BorrowedBook> BorrowedBooksWithISBN(String ISBN)
...
             if (borrowedBooksISBN==null||borrowedBooksISBN.Count==0)
             {
                 throw new ArgumentNullException(nameof(borrowedBooksISBN), "In DepartmentReaders.GetTimeWhenFreeBook");
             }
             DateTime timeNow = DateTime.Now;
-            BorrowedBook moreBestFreeBook = borrowedBooksISBN.OrderBy(book => book.dateTime).FirstOrDefault(book => (timeNow - book.dateTime).Days < 30);
-            return moreBestFreeBook.dateTime- timeNow;
+            int? dayFreeBook = borrowedBooksISBN.OrderBy(book => book.dateTime).FirstOrDefault(book => (timeNow - book.dateTime).Days < 30)?.dateTime.AddDays(30).Day;
+            if (dayFreeBook == null) return null;
+            return DateTime.Now.AddDays(dayFreeBook.Value);
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/DepartmentReaders.cs]: код стал еще сложнее. чтобы сделать его проще, тебе надо только посчитать дату возврата в момент выдачи книги. и сохранить её. а в этом методе ничег оне высчитывать, а просто отсортировать выданные книги по дате возврата.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary.Test/TestLibrary.cs (hunk 10 above)
Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary.Test/TestLibrary.cs]: 1. ну так вот ты и учишься тестировать без UI :)
2. дату можно не проверять в сообщении, смотри ShouldContain и остальные методы.

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 13 above)
Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: but the librarian have to make sure that the same book has returned

File: ClassLibrary/ClassLibrary/Library.cs (hunk 14 above)
Comment (alex-skiba-postnl) [reply ClassLibrary/ClassLibrary/Library.cs]: you can cover anything with tests :)

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs [hunk 16]
```diff
@@ -48,9 +48,21 @@ public int CountBorrowedBooksWithISBN(String ISBN)
...
+        public DateTime? GetDayWhenFreeBook(List<BorrowedBook> borrowedBooksISBN)
+        {
+            if (borrowedBooksISBN==null||borrowedBooksISBN.Count==0)
+            {
+                throw new ArgumentNullException(nameof(borrowedBooksISBN), "In DepartmentReaders.GetTimeWhenFreeBook");
+            }
+            DateTime timeNow = DateTime.Now;
+            DateTime? dayFreeBook = borrowedBooksISBN.OrderBy(book => book.dateTime).FirstOrDefault(book => (timeNow - book.dateTime).Days < 30)?.dateTime.AddDays(30);
+            if (dayFreeBook == null) return null;
+            return dayFreeBook;
```
Comment (alex-skiba-postnl) [inline ClassLibrary/ClassLibrary/DepartmentReaders.cs]: код стал еще сложнее. чтобы сделать его проще, тебе надо только посчитать дату возврата в момент выдачи книги. и сохранить её. а в этом методе ничег оне высчитывать, а просто отсортировать выданные книги по дате возврата.

**Review by alex-skiba-postnl — CHANGES_REQUESTED** [review]

File: ClassLibrary/ClassLibrary.Test/TestLibrary.cs (hunk 10 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary.Test/TestLibrary.cs]: +

File: ClassLibrary/ClassLibrary/Library.cs (hunk 14 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/Library.cs]: Test BookAvaible30Days

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 16 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: I did

File: ClassLibrary/ClassLibrary/DepartmentReaders.cs (hunk 13 above)
Comment (prajdziswet) [reply ClassLibrary/ClassLibrary/DepartmentReaders.cs]: не понял это был ответ или это было замечнаие добавить еще доп. проверку
